File: frontend/src/types/compare.ts
export type Confidence = "HIGH" | "MED" | "LOW" | "NONE";

export type TradeSide = {
  trade_id?: string | number;
  ts?: string | null;
  slot?: number | null;
  price?: number | null;
  roi_pct?: number | null;
  tip_lamports?: number | null;
  cu_used?: number | null;
  cu_price_micro_lamports?: number | null;
  tx_route?: string | null;
};

export type Deltas = {
  slots?: number | null;
  ms?: number | null;
  price_drift_pct?: number | null;
  roi_delta_pct?: number | null;
};

export type Diagnostics = {
  cause?: string | null;
  detail?: string | null;
};

export type ComparePayload = {
  pair_id?: string | null;
  side?: "BUY" | "SELL";
  token_mint?: string;
  copy: TradeSide;
  source: TradeSide;
  deltas: Deltas;
  diagnostics: Diagnostics;
  confidence: Confidence;
};

export type AwaitingPayload = {
  status: "AWAITING_MATCH";
  message: string;
  confidence: Confidence;
};

export type CompareResponse = ComparePayload | AwaitingPayload;

File: backend/app/schemas/compare.py
from pydantic import BaseModel, Field
from typing import Optional, Literal

Confidence = Literal["HIGH", "MED", "LOW", "NONE"]

class TradeSide(BaseModel):
    trade_id: Optional[str] = None
    ts: Optional[str] = None
    slot: Optional[int] = None
    price: Optional[float] = None
    roi_pct: Optional[float] = None
    tip_lamports: Optional[int] = None
    cu_used: Optional[int] = None
    cu_price_micro_lamports: Optional[int] = None
    tx_route: Optional[str] = None

class Deltas(BaseModel):
    slots: Optional[int] = None
    ms: Optional[int] = None
    price_drift_pct: Optional[float] = None
    roi_delta_pct: Optional[float] = None

class Diagnostics(BaseModel):
    cause: Optional[str] = None
    detail: Optional[str] = None

class ComparePayload(BaseModel):
    # Keep JSON the same, avoid name clash by using alias
    pair_id: Optional[str] = None
    side: Optional[str] = None
    token: Optional[str] = Field(None, alias="token_mint")

    copy_side: TradeSide = Field(alias="copy")
    source: TradeSide

    deltas: Deltas
    diagnostics: Diagnostics
    confidence: Confidence = "NONE"

    model_config = {
        # allow passing either field names or aliases
        "populate_by_name": True
    }

class AwaitingPayload(BaseModel):
    status: Literal["AWAITING_MATCH"] = "AWAITING_MATCH"
    message: str = "Awaiting match…"
    confidence: Confidence = "LOW"
